In [ ]:
# %pip install pandas

# Section-1 Pandas


## 1.1 Import & load

In [1]:
import pandas as pd

import the pandas library and give it the short name "pd"

In [4]:
df = pd.read_csv(r"C:\Work\Programing Language\Python\Project\AI ML PROJECT\IPL_Stats\Titanic-Dataset.csv")

read the CSV file from disk and store it as a DataFrame, a DataFrame is like an Excel spreadsheet — rows and columns , "df" is just a common variable name for DataFrames

## 1.2 First Look

In [5]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


* show the FIRST 5 rows of the dataset
- `.head()` is used to take a quick peek at what the data looks like
- you can pass a number too: `df.head(10)` shows 10 rows

In [7]:
print("Shape of Dataset : ",df.shape)

Shape of Dataset :  (891, 12)


* `.shape` gives ``(rows, columns)`` as a tuple
* e.g. (891, 12) means 891 passengers and 12 columns

In [9]:
print(df.dtypes)

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object


* `.dtypes` shows the data TYPE of each coloum
* `int64` = whole number
* `float` = decimals
* `object` = text/strings


In [10]:
print("Dataset Information : ",df.info())

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB
Dataset Information :  None


* `.info()` gives a full summary of the DataFrame
* shows column names, non-null counts, and data types
* very useful to spot missing data at a glance


In [11]:
print(df.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


* `.describe()` gives statistics for all numeric columns
* shows count, mean, std, min, 25%, 50%, 75%, max
* great for understanding the "spread" of your data

## Find Missing Values

In [12]:
df.isnull()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,True,False
887,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,False,False,True,False,False,False,False,True,False
889,False,False,False,False,False,False,False,False,False,False,False,False


* `.isnull()` returns a DataFrame of True/False
 * True = that cell is missing (NaN)
  * False = it has a value


In [13]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

* `.sum()` adds up all the Trues in each columnm
* gives total count of missing values per column
 * Age: 177 missing

## Cell 1.4 : Filling & Dropping Nulls

In [14]:
print("Before:", df.isnull().sum())

Before: PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


* printing before cleaning so that we can compare

In [15]:
df['Age'] = df['Age'].fillna(
    df['Age'].median()
)

* `.fillna()` replaces all NaN values with something you choose
 * we fill with the `MEDIAN` age — more robust than mean

In [16]:
df = df.drop(columns=['Cabin'])

* `.drop(columns=[...])` removes entire columns from the DataFrame
* we drop `Cabin` because `77%` of its values are missing
* keeping it would confuse any model we train later

In [17]:
df['Embarked'] = df['Embarked'].fillna(   # fill the 2 missing Embarked values
    df['Embarked'].mode()[0]              # .mode() returns the most common value (S, C, or Q)
)

* fill the 2 missing Embarked values
* `.mode()` returns most common values ( S,C,or Q)

In [18]:
print("After:", df.isnull().sum())

After: PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


* print missing counts AFTER cleaning to confirm everything is fixed

## 1.5 Selecting and Filtering

In [19]:
df['Survived']

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

* selecting a single column returns a Series (1D)
* 0 = did not survive, 1 = survived

In [20]:
df[['Name', 'Age', 'Survived']]

,Name,Age,Survived
0,"Braund, Mr. Owen Harris",22.0,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1
2,"Heikkinen, Miss. Laina",26.0,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1
4,"Allen, Mr. William Henry",35.0,0
...,...,...,...
886,"Montvila, Rev. Juozas",27.0,0
887,"Graham, Miss. Margaret Edith",19.0,1
888,"Johnston, Miss. Catherine Helen ""Carrie""",28.0,0
889,"Behr, Mr. Karl Howell",26.0,1


* selecting MULTIPLE columns using a list inside []
* always use double brackets `[[...]]` for multiple columns

In [22]:
survivors = df[df['Survived'] == 1]
print(survivors.shape)

(342, 11)


* boolean filtering: keep only rows where Survived equals 1
* `df['Survived'] == 1` creates a `True/False` Series
 * the outer `df[...]` keeps only the True rows